### PR TITLE
Extend FramingPolicy to include allowed sources.

### PR DESCRIPTION
**Pull request by mtrzos**
This is the first step to add a full support for framing protection (described in #77)

> It's a good idea to open an issue first for discussion.

- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

File: safehttp/plugins/csp/csp.go [hunk 1]
```diff
@@ -159,6 +162,15 @@ func (f FramingPolicy) Serialize(nonce string) string {
 	var b strings.Builder
 	b.WriteString("frame-ancestors 'self'")
```
Comment (kele) [inline safehttp/plugins/csp/csp.go:163]: At this point I'd carve out the `frame-ancestors` and `report-uri` directives to separate functions and join them with `; `.

e.g.:

```
ancestors := frameAncestors(f.Sources)
rep := report(f.ReportURI)

return strings.Join([]string{ancestors, rep}, "; ")
```

Alternatively:

```
var b strings.Builder
f.writeAncestors(&b)
f.writeReportURI(&b)
return b.String()
```

(the `writeFoo` methods always append `; `)

**Review by kele — CHANGES_REQUESTED** [review]

File: safehttp/plugins/csp/csp.go (hunk 1 above)
Comment (mtrzos) [reply safehttp/plugins/csp/csp.go:163]: I chose the first option. It adds a semicolon to the default policy but it's still valid.

File: safehttp/plugins/csp/csp.go [hunk 2]
```diff
@@ -159,9 +162,30 @@ func (f FramingPolicy) Serialize(nonce string) string {
...
 
-	if f.ReportURI != "" {
-		b.WriteString("; report-uri ")
-		b.WriteString(f.ReportURI)
+	ancestors := frameAncestors(f.Sources)
+	rep := report(f.ReportURI)
+	b.WriteString(strings.Join([]string{ancestors, rep}, "; "))
```
Comment (kele) [inline safehttp/plugins/csp/csp.go:167]: I'm a bit nitpicky here, but please bear with me.

As it is, you have four lines:
1. Use `WriteString` to write a part of the `frame-ancestors` directive.
2. Create a string that contains a part of the `frame-ancestors` directive.
3. Create a string that contains the full `report-uri` directive.
4. Use `WriteString` to write the two parts above, joined by `"; "`.

Wouldn't it be cleaner to merge 1 and 2 inside `frameAncestors`?

This way the code would end up being:
```
var b strings.Builder
b.WriteString(frameAncestors(f.Sources)) // writes at least "frame-ancestors 'self'; " (note the space)
b.WriteString(report(f.ReportURI)) // writes a "; " delimited directive or an empty string
return strings.TrimSpace(b.String())
```

**Review by kele — CHANGES_REQUESTED** [review]

File: safehttp/plugins/csp/csp.go (hunk 2 above)
Comment (mtrzos) [reply safehttp/plugins/csp/csp.go]: That's true. It's simpler and avoids extra unneeded semicolon in the case report-uri is not present.

File: safehttp/plugins/csp/csp.go [hunk 3]
```diff
@@ -157,11 +160,31 @@ type FramingPolicy struct {
...
+	}
+
+	return b.String()
+}
+
+func report(reportURI string) string {
+	var b strings.Builder
+
+	if reportURI != "" {
 		b.WriteString("; report-uri ")
```
Comment (kele) [inline safehttp/plugins/csp/csp.go:186]: This is brittle.

There is no reason for `Serialize` to not write `report()` first and then `frameAncestors`, but putting the semicolon at the beginning breaks this.

**Review by kele — APPROVED** [review]
One last nit. Otherwise LGTM.

File: safehttp/plugins/csp/csp.go (hunk 3 above)
Comment (mtrzos) [reply safehttp/plugins/csp/csp.go]: Ok, I made it order agnostic.

**Comment by kele** [comment]
Thank you! :)

**Pull request merged**